### PR TITLE
Align Receiver metrics to Eventing core Ingress metrics

### PR DESCRIPTION
**Pull request by pierDipi**
See https://knative.dev/docs/eventing/observability/metrics/eventing-metrics/#broker-ingress

_Note: Eventing Core uses `broker_name` for the tag but since we have Broker/Channel/Sink, I'm using `name` instead._

Example metrics:
```
# HELP event_count_total Number of events received
# TYPE event_count_total counter
event_count_total{event_type="com.example.FullEvent",name="",namespace_name="",response_code="202",response_code_class="2xx",} 15915.0

# HELP event_dispatch_latencies_max The time spent dispatching an event to Kafka
# TYPE event_dispatch_latencies_max gauge
event_dispatch_latencies_max{event_type="com.example.FullEvent",name="",namespace_name="",response_code="202",response_code_class="2xx",} 99.0
# HELP event_dispatch_latencies The time spent dispatching an event to Kafka
# TYPE event_dispatch_latencies histogram
event_dispatch_latencies_bucket{event_type="com.example.FullEvent",name="",namespace_name="",response_code="202",response_code_class="2xx",le="1.0",} 7499.0
event_dispatch_latencies_bucket{event_type="com.example.FullEvent",name="",namespace_name="",response_code="202",response_code_class="2xx",le="2.0",} 14566.0
event_dispatch_latencies_bucket{event_type="com.example.FullEvent",name="",namespace_name="",response_code="202",response_code_class="2xx",le="5.0",} 15742.0
event_dispatch_latencies_bucket{event_type="com.example.FullEvent",name="",namespace_name="",response_code="202",response_code_class="2xx",le="10.0",} 15862.0
event_dispatch_latencies_bucket{event_type="com.example.FullEvent",name="",namespace_name="",response_code="202",response_code_class="2xx",le="20.0",} 15894.0
event_dispatch_latencies_bucket{event_type="com.example.FullEvent",name="",namespace_name="",response_code="202",response_code_class="2xx",le="50.0",} 15907.0
event_dispatch_latencies_bucket{event_type="com.example.FullEvent",name="",namespace_name="",response_code="202",response_code_class="2xx",le="100.0",} 15915.0
event_dispatch_latencies_bucket{event_type="com.example.FullEvent",name="",namespace_name="",response_code="202",response_code_class="2xx",le="500.0",} 15915.0
event_dispatch_latencies_bucket{event_type="com.example.FullEvent",name="",namespace_name="",response_code="202",response_code_class="2xx",le="1000.0",} 15915.0
event_dispatch_latencies_bucket{event_type="com.example.FullEvent",name="",namespace_name="",response_code="202",response_code_class="2xx",le="5000.0",} 15915.0
event_dispatch_latencies_bucket{event_type="com.example.FullEvent",name="",namespace_name="",response_code="202",response_code_class="2xx",le="10000.0",} 15915.0
event_dispatch_latencies_bucket{event_type="com.example.FullEvent",name="",namespace_name="",response_code="202",response_code_class="2xx",le="+Inf",} 15915.0
event_dispatch_latencies_count{event_type="com.example.FullEvent",name="",namespace_name="",response_code="202",response_code_class="2xx",} 15915.0
event_dispatch_latencies_sum{event_type="com.example.FullEvent",name="",namespace_name="",response_code="202",response_code_class="2xx",} 27753.0
```
_Note: `name="",namespace_name=""` tags require control plane changes that will be done in a follow up PR (edit: done https://github.com/knative-sandbox/eventing-kafka-broker/pull/1701)._

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

### CODE REVIEW

**Comment by pierDipi** [comment]
Dumping the metrics from the eventing core ingress:
```
# HELP mt_broker_ingress_event_count Number of events received by a Broker
# TYPE mt_broker_ingress_event_count counter
mt_broker_ingress_event_count{broker_name="default",container_name="ingress",event_type="dev.knative.samples.helloworld",namespace_name="default",response_code="202",response_code_class="2xx",unique_name="mt-broker-ingress-685f5554c-shjba958bbb2c286bbe73021d54cdaba03b"} 1
# HELP mt_broker_ingress_event_dispatch_latencies The time spent dispatching an event to a Channel
# TYPE mt_broker_ingress_event_dispatch_latencies histogram
mt_broker_ingress_event_dispatch_latencies_bucket{broker_name="default",container_name="ingress",event_type="dev.knative.samples.helloworld",namespace_name="default",response_code="202",response_code_class="2xx",unique_name="mt-broker-ingress-685f5554c-shjba958bbb2c286bbe73021d54cdaba03b",le="1"} 0
mt_broker_ingress_event_dispatch_latencies_bucket{broker_name="default",container_name="ingress",event_type="dev.knative.samples.helloworld",namespace_name="default",response_code="202",response_code_class="2xx",unique_name="mt-broker-ingress-685f5554c-shjba958bbb2c286bbe73021d54cdaba03b",le="2"} 1
mt_broker_ingress_event_dispatch_latencies_bucket{broker_name="default",container_name="ingress",event_type="dev.knative.samples.helloworld",namespace_name="default",response_code="202",response_code_class="2xx",unique_name="mt-broker-ingress-685f5554c-shjba958bbb2c286bbe73021d54cdaba03b",le="5"} 1
mt_broker_ingress_event_dispatch_latencies_bucket{broker_name="default",container_name="ingress",event_type="dev.knative.samples.helloworld",namespace_name="default",response_code="202",response_code_class="2xx",unique_name="mt-broker-ingress-685f5554c-shjba958bbb2c286bbe73021d54cdaba03b",le="10"} 1
mt_broker_ingress_event_dispatch_latencies_bucket{broker_name="default",container_name="ingress",event_type="dev.knative.samples.helloworld",namespace_name="default",response_code="202",response_code_class="2xx",unique_name="mt-broker-ingress-685f5554c-shjba958bbb2c286bbe73021d54cdaba03b",le="20"} 1
mt_broker_ingress_event_dispatch_latencies_bucket{broker_name="default",container_name="ingress",event_type="dev.knative.samples.helloworld",namespace_name="default",response_code="202",response_code_class="2xx",unique_name="mt-broker-ingress-685f5554c-shjba958bbb2c286bbe73021d54cdaba03b",le="50"} 1
mt_broker_ingress_event_dispatch_latencies_bucket{broker_name="default",container_name="ingress",event_type="dev.knative.samples.helloworld",namespace_name="default",response_code="202",response_code_class="2xx",unique_name="mt-broker-ingress-685f5554c-shjba958bbb2c286bbe73021d54cdaba03b",le="100"} 1
mt_broker_ingress_event_dispatch_latencies_bucket{broker_name="default",container_name="ingress",event_type="dev.knative.samples.helloworld",namespace_name="default",response_code="202",response_code_class="2xx",unique_name="mt-broker-ingress-685f5554c-shjba958bbb2c286bbe73021d54cdaba03b",le="200"} 1
mt_broker_ingress_event_dispatch_latencies_bucket{broker_name="default",container_name="ingress",event_type="dev.knative.samples.helloworld",namespace_name="default",response_code="202",response_code_class="2xx",unique_name="mt-broker-ingress-685f5554c-shjba958bbb2c286bbe73021d54cdaba03b",le="500"} 1
mt_broker_ingress_event_dispatch_latencies_bucket{broker_name="default",container_name="ingress",event_type="dev.knative.samples.helloworld",namespace_name="default",response_code="202",response_code_class="2xx",unique_name="mt-broker-ingress-685f5554c-shjba958bbb2c286bbe73021d54cdaba03b",le="1000"} 1
mt_broker_ingress_event_dispatch_latencies_bucket{broker_name="default",container_name="ingress",event_type="dev.knative.samples.helloworld",namespace_name="default",response_code="202",response_code_class="2xx",unique_name="mt-broker-ingress-685f5554c-shjba958bbb2c286bbe73021d54cdaba03b",le="2000"} 1
mt_broker_ingress_event_dispatch_latencies_bucket{broker_name="default",container_name="ingress",event_type="dev.knative.samples.helloworld",namespace_name="default",response_code="202",response_code_class="2xx",unique_name="mt-broker-ingress-685f5554c-shjba958bbb2c286bbe73021d54cdaba03b",le="5000"} 1
mt_broker_ingress_event_dispatch_latencies_bucket{broker_name="default",container_name="ingress",event_type="dev.knative.samples.helloworld",namespace_name="default",response_code="202",response_code_class="2xx",unique_name="mt-broker-ingress-685f5554c-shjba958bbb2c286bbe73021d54cdaba03b",le="10000"} 1
mt_broker_ingress_event_dispatch_latencies_bucket{broker_name="default",container_name="ingress",event_type="dev.knative.samples.helloworld",namespace_name="default",response_code="202",response_code_class="2xx",unique_name="mt-broker-ingress-685f5554c-shjba958bbb2c286bbe73021d54cdaba03b",le="+Inf"} 1
mt_broker_ingress_event_dispatch_latencies_sum{broker_name="default",container_name="ingress",event_type="dev.knative.samples.helloworld",namespace_name="default",response_code="202",response_code_class="2xx",unique_name="mt-broker-ingress-685f5554c-shjba958bbb2c286bbe73021d54cdaba03b"} 1
mt_broker_ingress_event_dispatch_latencies_count{broker_name="default",container_name="ingress",event_type="dev.knative.samples.helloworld",namespace_name="default",response_code="202",response_code_class="2xx",unique_name="mt-broker-ingress-685f5554c-shjba958bbb2c286bbe73021d54cdaba03b"} 1
```

**Comment by pierDipi** [comment]
FYI @skonto

**Comment by pierDipi** [comment]
This is good to go now!

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/IngressProducer.java [hunk 1]
```diff
@@ -45,4 +46,9 @@
    */
   String getTopic();
 
+  /**
+   * @return the resource associated with this producer.
+   */
+  DataPlaneContract.Reference getReference();
```
Comment (matzew) [inline data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/IngressProducer.java:52]: did we do the references on a different PR before?

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/RequestContext.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package dev.knative.eventing.kafka.broker.receiver;
+
+import io.vertx.core.http.HttpServerRequest;
+
+public class RequestContext {
```
Comment (matzew) [inline data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/RequestContext.java:20]: is this wrapper really a "generic" context? IMO sounds broader than it is, since we just calculate the latency as an additional thing

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/impl/handler/IngressRequestHandlerImpl.java [hunk 3]
```diff
@@ -43,36 +46,58 @@
...
+                                   final MeterRegistry meterRegistry) {
     this.requestToRecordMapper = requestToRecordMapper;
-    this.badRequestCounter = badRequestCounter;
-    this.produceEventsCounter = produceEventsCounter;
+    this.meterRegistry = meterRegistry;
   }
 
   @Override
-  public void handle(HttpServerRequest request, IngressProducer producer) {
+  public void handle(final RequestContext requestContext, final IngressProducer producer) {
```
Comment (matzew) [inline data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/impl/handler/IngressRequestHandlerImpl.java:82]: I like the `final` fetish here! Good stuff !!!

**Review by matzew — APPROVED** [review]
/lgtm
/approve

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/RequestContext.java (hunk 2 above)
Comment (pierDipi) [reply data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/RequestContext.java:20]: Are you suggesting to call it `RequestWrapper` rather than `RequestContext`?

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/IngressProducer.java (hunk 1 above)
Comment (pierDipi) [reply data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/IngressProducer.java:52]: Yes, I referenced the PR in the PR body: https://github.com/knative-sandbox/eventing-kafka-broker/pull/1701

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/RequestContext.java (hunk 2 above)
Comment (matzew) [reply data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/RequestContext.java:20]: I am not sure - but I am also fine w/ the `RequestContext` (I do find naming hard, tbh)

**Review by matzew — APPROVED** [review]
/lgtm
/approve

**Comment by skonto** [comment]
Let me verify it will report back.

**Comment by skonto** [comment]
@pierDipi we might want to use `name` from now on. Eventing sources use `name` as well.

**Comment by pierDipi** [comment]
> @pierDipi we might want to use `name` from now on. Eventing sources use `name` as well. `namespace_name` is meant for the ns that the initial resource is created in. For this case I guess it is the receiver ns correct?

I'm setting `namespace_name` to the resource namespace `metadata.namespace`, the receiver is deployed in the system namespace `knative-eventing` and it's for every resource
